### DISCUSSION TITLE
Change schema after GraphQL service started

### DISCUSSION THREAD

**Discussion by t0ry** [graphql-java/graphql-java#3074] [General]
So I have spring-boot graph-ql based service. I have 95% dynamic data model in my system which is defined by customer via specific API. Thus my schema in turn is defined by outside source. Moreover the data model might be modified at any point of time without any restrictions in frequency (let's assume generally once or twice per week). As a consequence schema must be up to date. Let's say I have events based mechanism to notify the service about changes.

Is it even possible to build schema which might be changed at any time after service started? 
I looked at schema transferring here: https://www.graphql-java.com/documentation/schema/#changing-schema
But I can't get how can I substitute existing schema with new one while service already running?
`SchemaTransformer.transformSchema(existingSchema, visitor)` returns new schema but how to put it instead the old one? 

Thanks in advance.

**Comment by bbakerman** [comment]
Sorry for the late reply.   I have been on a long vacation.

Your right that `SchemaTransformer` is the way to change to schema definition along with `graphql.schema.GraphQLCodeRegistry` for the data fetchers.

Your question though really is a Spring Graphql question and I am not the best person to answer it.  The `graphql.schema.GraphQLSchema` is injected as a bean when the Spring application context starts.  This is done via the spring-graphql code.

Somehow you need to have that bean change dynamically at runtime.  This is part I don't understand yet.  I am going to do some digging to find out more.  I think you can use a customizer to adjust it "as the context is built" however that is a one time spring application context thing.  How you would dynamic change this bean at any time and make the spring context aware of that is what I don't know.

**Reply by bbakerman** [reply]
It would be worth asking in https://github.com/spring-projects/spring-graphql/issues as well

**Reply by t0ry** [reply]
bbakerman,
thanks for reply. After digging into the code I have an understanding similar to what you described.

**Reply by bbakerman** [reply]
FYI - Over the years we have gotten this request a lot - that is I have some metadata (somewhere) that causes the schema to change without a application re-deploy.

A gateway might require this but some people love to build “contract driven” systems.

So the ability to “re-create” the schema (according to some external event - hand waving here) is common enough

We certainly do this in our Atlassian Graphql Gateway (nadel based) that I work on.

One of the challenges of this is of course

 * what do you do in a N server world
    * your schema will never be atomically changed
  * what do you do with in flight requests
    * a request in flight really should use the same schema config as it started with
    * so the new schema needs to be carefully put in place
    * we do this in the Atlassian Gateway by
        * setting up a “schema instance” per request - eg kinda give it an id
        * when we ask for the schema, we do it be map look up by id
        * the new schema is given out to future requests
        * the old schema is held in memory for say 2 minutes
        * after 2 minute the requests are considered all done
        * the old schema can then be thrown away

**Comment by bbakerman** [comment]
I had a look into this and also spoke to some of the Spring developers about this.

They key interface is `org.springframework.graphql.execution.GraphQlSource` - this is the bean that gives out the schema and the `graphql.GraphQL` that holds it.

Each request ends up calling `org.springframework.graphql.execution.GraphQlSource#graphQl` to execute the query.  

So in theory of you had your own `GraphQlSource` bean - then you could could dynamically give out your own dynamic schema and associated `graphql.GraphQL`.

The challenge here is that Spring does a fair bit of work in building the auto configured `GraphQlSource` 

See `org.springframework.boot.autoconfigure.graphql.GraphQlAutoConfiguration#graphQlSource` for  a bunch of common code that finds schema files, allows customisations, adds instrumentations and so on.  You need to repeat much of that code I suspect to get your own `GraphQlSource` in place

**Reply by bbakerman** [reply]
```
@Bean
	@ConditionalOnMissingBean
	public GraphQlSource graphQlSource(ResourcePatternResolver resourcePatternResolver, GraphQlProperties properties,
			ObjectProvider<DataFetcherExceptionResolver> exceptionResolvers,
			ObjectProvider<SubscriptionExceptionResolver> subscriptionExceptionResolvers,
			ObjectProvider<Instrumentation> instrumentations, ObjectProvider<RuntimeWiringConfigurer> wiringConfigurers,
			ObjectProvider<GraphQlSourceBuilderCustomizer> sourceCustomizers) {
		String[] schemaLocations = properties.getSchema().getLocations();
		Resource[] schemaResources = resolveSchemaResources(resourcePatternResolver, schemaLocations,
				properties.getSchema().getFileExtensions());
		GraphQlSource.SchemaResourceBuilder builder = GraphQlSource.schemaResourceBuilder()
				.schemaResources(schemaResources).exceptionResolvers(exceptionResolvers.orderedStream().toList())
				.subscriptionExceptionResolvers(subscriptionExceptionResolvers.orderedStream().toList())
				.instrumentation(instrumentations.orderedStream().toList());
		if (!properties.getSchema().getIntrospection().isEnabled()) {
			builder.configureRuntimeWiring(this::enableIntrospection);
		}
		wiringConfigurers.orderedStream().forEach(builder::configureRuntimeWiring);
		sourceCustomizers.orderedStream().forEach((customizer) -> customizer.customize(builder));
		return builder.build();
	}
```

There is lot going on here

**Reply by bbakerman** [reply]
This code kinda works

```
@Component
public class OurBeanPostProcessor implements BeanPostProcessor {
    @Override
    public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
        if (bean instanceof GraphQlSource original) {
            return new GraphQlSource() {
                @Override
                public GraphQL graphQl() {
                    return original.graphQl();
                }

                @Override
                public GraphQLSchema schema() {
                    return original.schema();
                }
            };
        }
        return bean;
    }
}
```

That is you can post process to Spring built `GraphQlSource` and hence give back one that is dynamic.

I am unsure if this is good spring practice or not (I am asking some other Spring devs on this).  But it would allow Spring to builds its standard `GraphQlSource` and then you could dynamically change it later without application restart.

The problems I mention above about in flight requests will not be solved however.  This would be a hard "swap" and hence its possible an inflight requests could get Schema A and then get Schema B on some other part of call chain.

**Reply by bbakerman** [reply]
I asked some Spring devs about `postProcessAfterInitialization` - they gave it the 👍
